prov/verbs/src/verbs_info.c

```diff
@@ -885,29 +886,45 @@ int fi_ibv_get_srcaddr_devs(struct fi_info *info)
 		return 0;
 	}
 
-	for (fi = info; fi; fi = fi->next) {
+	for (fi = *info; fi;) {
+		dev_found = 0;
 		dlist_foreach_container(&verbs_devs, dev, entry)
-			if (!strncmp(fi->domain_attr->name, dev->name, strlen(dev->name)))
+			if (!strncmp(fi->domain_attr->name, dev->name, strlen(dev->name))) {
+				dev_found = 1;
 				break;
+			}
 
-		dlist_foreach_container(&dev->addrs, addr, entry) {
-			/* When a device has multiple interfaces/addresses configured
-			 * duplicate fi_info and add the address info. fi->src_addr
-			 * would have been set in the previous iteration */
-			if (fi->src_addr) {
-				if (!(add_info = fi_dupinfo(fi))) {
-					ret = -FI_ENOMEM;
-					goto out;
+		if(dev_found) {
+			dlist_foreach_container(&dev->addrs, addr, entry) {
+				/* When a device has multiple interfaces/addresses configured
+				 * duplicate fi_info and add the address info. fi->src_addr
+				 * would have been set in the previous iteration */
+				if (fi->src_addr) {
+					if (!(add_info = fi_dupinfo(fi))) {
+						ret = -FI_ENOMEM;
+						goto out;
+					}
+
+					add_info->next = fi->next;
+					fi->next = add_info;
+					fi = add_info;
 				}
 
-				add_info->next = fi->next;
-				fi->next = add_info;
-				fi = add_info;
+				ret = fi_ibv_rai_to_fi(addr->rai, fi);
+				if (ret)
+					goto out;
 			}
-
-			ret = fi_ibv_rai_to_fi(addr->rai, fi);
-			if (ret)
-				goto out;
+			fi = fi->next;
+		} else { /* src_addr required to bind listener. exclude info
+			    without src_addr.
+			    this situation may happen when interface is up,
+			    but not configured (no IP address) */
+			if(fi == *info)
+				*info = fi->next;
+			to_remove = fi;
+			fi = fi->next;
+			to_remove->next = NULL;
+			fi_freeinfo(to_remove);
```

Comment:
the most common way to use libfabric is get fi_info based on hints & use first element. In case if we leave such incorrect records application may fail to initialize (we have such situation in IMPI).

what about shift such elements to tail of list?